### PR TITLE
Refactor: Extract say to its own file

### PR DESCRIPTION
**Pull request by gregorriegler**
A first step tackling #296

### CODE REVIEW

**Comment by hollesse** [comment]
I really like the idea, but I would have expected it to be its own package and export just few functions and not all. What do you think about this?

File: say.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func sayFix(instruction string, command string) {
+	sayWithPrefix(instruction, "👉 ")
+	sayEmptyLine()
+	sayIndented(command)
+	sayEmptyLine()
+}
+
+func sayNext(instruction string, command string) {
```
Comment (hollesse) [inline say.go:47]: then this should be public

File: say.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	sayWithPrefix(instruction, "👉 ")
+	sayEmptyLine()
+	sayIndented(command)
+	sayEmptyLine()
+}
+
+func sayWithPrefix(s string, prefix string) {
```
Comment (hollesse) [inline say.go:54]: here I am not sure if this should be public or private

File: say.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func say(s string) {
+	if len(s) == 0 {
+		return
+	}
+	printToConsole(strings.TrimRight(s, " \r\n\t\v\f") + "\n")
+}
+
+func debugInfo(text string) {
```
Comment (hollesse) [inline say.go:72]: then this should be public

File: say.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	if Debug {
+		sayWithPrefix(text, "DEBUG ")
+	}
+}
+
+var printToConsole = func(message string) {
```
Comment (hollesse) [inline say.go:78]: why is this an anonymous function?

**Review by hollesse — CHANGES_REQUESTED** [review]
I think we should use a different package and have some functions private

**Comment by gregorriegler** [comment]
You're just teaching me some go, and I love it!

File: say.go (hunk 4 above)
Comment (gregorriegler) [reply say.go:78]: tried making it a func. found out that its overridden in mob_test so leaving it for now.

**Comment by gregorriegler** [comment]
I'm stuck. The tests finish with an exit code and I don't know why.

**Comment by gregorriegler** [comment]
Ok, I understand now. Trying to find a solution

**Comment by gregorriegler** [comment]
solved

File: coauthors.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package main
 import (
 	"bufio"
 	"fmt"
+	"github.com/remotemobprogramming/mob/v3/say"
```
Comment (hollesse) [inline coauthors.go:6]: I'am not sure if this is the way to go. But disclaimer I also didn't knew that before I just looked up if there is something like private. For me this looks a bit strange but maybe I do not know how it correctly works.

File: say/say.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func WithPrefix(s string, prefix string) {
+	lines := strings.Split(strings.TrimSpace(s), "\n")
+	for i := 0; i < len(lines); i++ {
+		PrintToConsole(prefix + strings.TrimSpace(lines[i]) + "\n")
+	}
+}
+
+func EmptyLine() {
```
Comment (hollesse) [inline say/say.go:65]: I think this could be private, so it should be `func emptyLine() {`

File: say/say.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	PrintToConsole("\n")
+}
+
+func Say(s string) {
```
Comment (hollesse) [inline say/say.go:69]: I saw this is used, but should we really use it outside of this package? I am not sure...

File: say/say_test.go [hunk 8]
```diff
@@ -0,0 +1 @@
+package say
```
Comment (hollesse) [inline say/say_test.go:1]: Do we need that file if we do not have tests for that?

File: squash_wip.go [hunk 9]
```diff
@@ -69,7 +70,7 @@ func getEnvGitEditor() (gitEditor string, gitSequenceEditor string) {
 func mobExecutable() string {
 	if isTestEnvironment() {
 		wd, _ := os.Getwd()
-		return "go run $(ls -1 " + wd + "/*.go | grep -v _test.go)"
+		return "cd " + wd + " && go run ."
```
Comment (hollesse) [inline squash_wip.go:73]: wow i did not expect that :D

File: say/say.go (hunk 7 above)
Comment (gregorriegler) [reply say/say.go:69]: Its used a lot it seems. could rename it, but I think we need it.

File: say/say_test.go (hunk 8 above)
Comment (gregorriegler) [reply say/say_test.go:1]: at first it asked me for it. but i will try and delete it again

File: coauthors.go (hunk 5 above)
Comment (hollesse) [reply coauthors.go:6]: I just read a bit about packages. This should be fine :)

File: say/say.go (hunk 7 above)
Comment (hollesse) [reply say/say.go:69]: The name is fine but I was just thinking if this is something more general which should not be used. But for now it's fine to just leave it as it is.

File: say/say_test.go (hunk 8 above)
Comment (hollesse) [reply say/say_test.go]: If we need to keep it we could also just write some tests ;)

**Comment by gregorriegler** [comment]
ok, we got this.

File: mob.go [hunk 10]
```diff
@@ -664,55 +654,55 @@ func printDeprecatedDoneSquashMessage(value string) {
 
 	if unquotedValue == "true" || unquotedValue == "false" {
 		newValue := doneSquash(unquotedValue)
-		say("MOB_DONE_SQUASH is set to the deprecated value " + value + ". Use the value " + newValue + " instead.")
+		say.Say("MOB_DONE_SQUASH is set to the deprecated value " + value + ". Use the value " + newValue + " instead.")
```
Comment (hollesse) [inline mob.go:657]: Shouldn't this be a Warning?

File: mob.go [hunk 11]
```diff
@@ -664,55 +654,55 @@ func printDeprecatedDoneSquashMessage(value string) {
...
 	}
 }
 
 func removed(key string, message string) {
 	if _, set := os.LookupEnv(key); set {
-		say("Configuration option '" + key + "' is no longer used.")
-		say(message)
+		say.Say("Configuration option '" + key + "' is no longer used.")
```
Comment (hollesse) [inline mob.go:663]: This can also be a Warning

File: mob.go [hunk 12]
```diff
@@ -664,55 +654,55 @@ func printDeprecatedDoneSquashMessage(value string) {
...
+		say.Say(message)
 	}
 }
 
 func deprecated(key string, message string) {
 	if _, set := os.LookupEnv(key); set {
-		say("Configuration option '" + key + "' is deprecated.")
-		say(message)
+		say.Say("Configuration option '" + key + "' is deprecated.")
+		say.Say(message)
```
Comment (hollesse) [inline mob.go:671]: same here

File: mob.go [hunk 13]
```diff
@@ -664,55 +654,55 @@ func printDeprecatedDoneSquashMessage(value string) {
...
 	}
 }
 
 func experimental(key string) {
 	if _, set := os.LookupEnv(key); set {
-		say("Configuration option '" + key + "' is experimental. Be prepared that this option will be removed!")
+		say.Say("Configuration option '" + key + "' is experimental. Be prepared that this option will be removed!")
```
Comment (hollesse) [inline mob.go:677]: same here

File: mob.go [hunk 14]
```diff
@@ -877,10 +867,10 @@ func (branch Branch) isOrphanWipBranch(configuration Configuration) bool {
 }
 
 func branch(configuration Configuration) {
-	say(silentgit("branch", "--list", "--remote", newBranch("*").addWipPrefix(configuration).remote(configuration).Name))
+	say.Say(silentgit("branch", "--list", "--remote", newBranch("*").addWipPrefix(configuration).remote(configuration).Name))
```
Comment (hollesse) [inline mob.go:870]: does that make sense? why don't we use `git` instead of `say.Say(silentgit(...))`

File: mob.go [hunk 15]
```diff
@@ -877,10 +867,10 @@ func (branch Branch) isOrphanWipBranch(configuration Configuration) bool {
...
 
 	// DEPRECATED
-	say(silentgit("branch", "--list", "--remote", newBranch("mob-session").remote(configuration).Name))
+	say.Say(silentgit("branch", "--list", "--remote", newBranch("mob-session").remote(configuration).Name))
```
Comment (hollesse) [inline mob.go:873]: same here

**Review by hollesse — APPROVED** [review]

**Comment by gregorriegler** [comment]
I agree with all the points you made, however it is not a part of the change I'm trying to make in this pr. I just want to move say to its own package. We should improve how say is used, but that's another step. Let's make small steps and integrate often.

**Comment by hollesse** [comment]
I agree. That's why I approved the PR but leaved the comments.

**Comment by gregorriegler** [comment]
Ah, I didn't get that. So I'll merge it then.

**Pull request merged**